server/src/proxyManagement.mjs

```diff
@@ -0,0 +1,191 @@
+        }
+
+        ws.on('message', function message(data) {
+            var buf = Buffer.from(data);
+            const oResponseMessage = {
+              jsonrpc: '2.0',
+              id: JSON.parse(buf.toString()).id,
+              ...JSON.parse(buf.toString())
+            };
+            const responseMessage = JSON.stringify(oResponseMessage);
+            receiverWSClient.send(responseMessage);
+            console.log(`Sent message to ripple: ${responseMessage} \n`);
+        });
+
+        receiverWSClient.on('error', (error) => {
+            console.log("receiver error: ", error)
+        })
+      })
+    } else {
+      throw new Error("Cannot establish websocket connection. Option \"url\" with ws://host:port or wss://host:port required")
+    }
+}
+
+function getProxyWSConnection() {
+    return websocketConnection
+}
+
+function sendRequest(payload) {
+    const ws = getProxyWSConnection()
+    if (!ws) {
+        throw new Error("websocketConnection not established")
+    }
+    waitForSocketConnection(ws, async function(socket){
+        console.log("Request sent to device: ", payload);
+        await socket.send(payload);
+        await new Promise(r => setTimeout(r, 2000));
+    });
+}
+
+// Make the function wait until the connection is made...
+function waitForSocketConnection(socket, callback){
+    setTimeout(
+        function () {
+            if (socket.readyState === 1) {
+                if (callback != null){
+                    callback(socket);
+                }
+            } else {
+                if(getProxyWSConnection()) {
+                    socket = getProxyWSConnection()
+                }
+                console.log("Wait for connection...")
+                waitForSocketConnection(socket, callback);
+            }
+        }, 1000); // wait 1 second for the connection...
+}
+
+function buildWSUrl() {
+    return new Promise(resolve => {
+        //support wss and ws
+        const wsUrlProtocol = 'ws://'
+        const port = 9998
+        const path = '/jsonrpc'
+        const deviceHost = process.env.thunderIP
+        resolve([
+            wsUrlProtocol,
+            deviceHost,
+            ':' + port,
+            path,
+            process.env.wsToken ? '?token=' + process.env.wsToken : null,
+        ].join(''))
+    })    
+}
+
+function close() {
+  if(websocketConnection) {
+    websocketConnection.close()
+    websocketConnection = null
+  }
+}
+
+// Return thunder token from actual device for ws connection
+function getThunderToken(request) {
+    return new Promise(async (resolve, reject) => {
+      let output = {
+        stdout: '',
+        stderr: '',
+      };
+      // If token already exists, return token
+      if(process.env.wsToken) {
+        output.stdout = process.env.wsToken
+        resolve(output)
+      }
+      const { query } = parse(request.url);
+      if(query && query.includes("token=") && query.length > 6) {
+        resolve(true)
```

Comment:
This is not resolving with a token the way the other "resolve()" calls are. Is this on purpose?

Reply:
Addressed